metric_providers/psu/energy/ac/ipmi/machine/provider.py

```diff
@@ -43,3 +44,19 @@ def read_metrics(self, run_id, containers=None):
         df = df.drop(columns='interval')  # clean up
 
         return df
+
+    def check_system(self):
+        # Run 'sensors' command and capture the output
+        ps = subprocess.run(['sudo', '/usr/sbin/ipmi-dcmi', '--get-system-power-statistics'], capture_output=True, text=True, check=False)
+        if ps.returncode != 0:
+            raise MetricProviderConfigurationError(f"\n\n{self._metric_name} provider could not be started.\nCannot run the 'sudo /usr/sbin/ipmi-dcmi --get-system-power-statistics' command.\n\nAre you running in a VM / cloud / shared hosting?\nIf so please disable the {self._metric_name} provider in the config.yml")
+
+        current_power_line = [line for line in ps.stdout.split('\n') if line.startswith('Current Power')]
```

Comment:
While your implementation is definitely faster, I wouldn't go so far to say that mine is crazy slow, especially if we're comparing with regexes. This code is only run once per GMT run on a small text snippet, not 1_000_000 times.

Though in the interest of our work here, hopefully this code will be run 1_000_000 times over its "lifetime" across many machines - so ok yes in that context I can see the point. Still, I don't think running it a million times in sequence like this gives us an accurate representation...

I re-ran the benchmark (adding your string function as well), but with num_iterations = 10

```
/tmp via 🐍 v3.10.12 
❯ py3 time.py 
Regex method time: 0.000164 seconds
String method (dan) time: 0.000023 seconds
String method (arne) time: 0.000008 seconds

/tmp via 🐍 v3.10.12 
❯ py3 time.py 
Regex method time: 0.000243 seconds
String method (dan) time: 0.000032 seconds
String method (arne) time: 0.000009 seconds

/tmp via 🐍 v3.10.12 
❯ py3 time.py 
Regex method time: 0.000163 seconds
String method (dan) time: 0.000024 seconds
String method (arne) time: 0.000009 seconds

/tmp via 🐍 v3.10.12 
❯ py3 time.py 
Regex method time: 0.000172 seconds
String method (dan) time: 0.000021 seconds
String method (arne) time: 0.000008 seconds
```
Running this many times, averaging, and then scaling up the result I think would be a much more fair comparision. Its actually interesting to me that the string methods scale that much worse than regexe's when batched together, that's not the result I would have expected! I originally did this because I was curious what the "real world" time cost of one run was, I didn't expect to see this sort of flipped scaling...

Regardless, your version is definitely much more efficent, its 2-3 times as fast.

Reply:
But now that I've rambled about this too, yes I agree it makes sense to put it in the script directly

Reply:
Done  - 
as a note, because of set -euo pipefail, if ipmi is missing or the sudo /usr/sbin/ipmi-dcmi --get-system-power-statistics command fails for any reason, it will already exit and spit into stderr (which is what I see on my machine), so that's why there's no explicit check for that.

Also this check would be good to have a look at on a machine that gives back real data. I tested the check function with:

```
if [ "$check_system" = True ]; then
    test="Current Power                        : 0 Watts
Minimum Power over sampling duration : 10 watts
Maximum Power over sampling duration : 0 watts
Average Power over sampling duration : 0 watts
Time Stamp                           : 12/15/2023 - 12:41:53
Statistics reporting time period     : 5000 milliseconds
Power Measurement                    : Active"

    #first_line=$(sudo /usr/sbin/ipmi-dcmi --get-system-power-statistics | head -1)
    first_line=$(echo $test | head -1)
    if ! [[ "$first_line" =~ ^"Current Power" ]]; then
        echo "Unable to find 'Current Power' in the output of 'sudo /usr/sbin/ipmi-dcmi --get-system-power-statistics' command. Found $first_line instead">&2
        exit 1
    fi

    current_power_watts=$(echo $first_line |  awk '{print $4}')
    if [[ $current_power_watts -le 0 ]]; then
        echo "Current Power Watts is $current_power_watts, which is unexpected">&2
        exit 1
    fi
    exit 0
fi
```
and various other test strings, but that's not a standin for real data from the ipmi invocation. Shell scripting can be finicky with these things